### PR TITLE
Parallel write option for RichRDDByteArray.saveFromByteArrays

### PR DESCRIPTION
**Pull request by jigold**
Needed to address #2245 and for export bgen (future PR).

### CODE REVIEW

File: src/main/scala/is/hail/utils/TempDir.scala [hunk 1]
```diff
@@ -38,7 +38,7 @@ object TempDir {
       createTempDir("/tmp", hConf))
 }
 
-class TempDir(localTempDir: String, tempDir: String) {
+case class TempDir(localTempDir: String, tempDir: String) {
```
Comment (danking) [inline src/main/scala/is/hail/utils/TempDir.scala]: What is the reason for this change?

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala [hunk 2]
```diff
@@ -8,30 +8,48 @@ import org.apache.spark.rdd.RDD
...
 class RichRDDByteArray(val r: RDD[Array[Byte]]) extends AnyVal {
-  def saveFromByteArrays(filename: String, tmpDir: String, header: Option[Array[Byte]] = None, deleteTmpFiles: Boolean = true) {
+  def saveFromByteArrays(filename: String, tmpDir: String, header: Option[Array[Byte]] = None,
+    deleteTmpFiles: Boolean = true, parallelWrite: Boolean = false) {
     val nullWritableClassTag = implicitly[ClassTag[NullWritable]]
     val bytesClassTag = implicitly[ClassTag[BytesOnlyWritable]]
     val hConf = r.sparkContext.hadoopConfiguration
 
-    val tmpFileName = hConf.getTemporaryFile(tmpDir)
+    hConf.delete(filename, recursive = true) // overwriting by default
```
Comment (danking) [inline src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: In VDS write we normally throw an error when writing to an existing file. Why is overwriting the right default here?

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala [hunk 3]
```diff
@@ -8,30 +8,48 @@ import org.apache.spark.rdd.RDD
...
       iter.map { bb =>
         bw.set(new BytesWritable(bb))
         (NullWritable.get(), bw)
       }
     }
 
-    RDD.rddToPairRDDFunctions(rMapped)(nullWritableClassTag, bytesClassTag, null)
-      .saveAsHadoopFile[ByteArrayOutputFormat](tmpFileName)
+    RDD.rddToPairRDDFunctions(rMappedWithHeader)(nullWritableClassTag, bytesClassTag, null)
+      .saveAsHadoopFile[ByteArrayOutputFormat](parallelOutputPath)
```
Comment (danking) [inline src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: The explicit `rddToPairRDDFunctions` is unnecessary, because it is implicit. Let's clean up this code while we're in here:
```scala
rMappedWithHeader.saveAsHadoopFile(parallelOutputPath)
```

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala [hunk 4]
```diff
@@ -8,30 +8,48 @@ import org.apache.spark.rdd.RDD
...
-    header.foreach { str =>
-      hConf.writeDataFile(tmpFileName + "/header") { s =>
-        s.write(str)
-      }
+    if (!hConf.exists(parallelOutputPath + "/_SUCCESS"))
+      fatal("write failed: no success indicator found")
+
+    if (!parallelWrite) {
+      val expNumPart = if (r.partitions.length == 0) 1 else r.getNumPartitions
+      hConf.copyMerge(parallelOutputPath, filename, expNumPart, deleteTmpFiles, hasHeader = false)
```
Comment (danking) [inline src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: `hasHeader = false` because we only added it to the first partition, right?

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala [hunk 5]
```diff
@@ -8,30 +8,48 @@ import org.apache.spark.rdd.RDD
...
 
-    header.foreach { str =>
-      hConf.writeDataFile(tmpFileName + "/header") { s =>
-        s.write(str)
-      }
+    if (!hConf.exists(parallelOutputPath + "/_SUCCESS"))
+      fatal("write failed: no success indicator found")
+
+    if (!parallelWrite) {
+      val expNumPart = if (r.partitions.length == 0) 1 else r.getNumPartitions
```
Comment (danking) [inline src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: is this equivalent to `rWithHeader.partitions.length`?

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala (hunk 3 above)
Comment (danking) [reply src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: I'm not sure if the type parameter is required or not, it seems that it might be required.

File: src/test/scala/is/hail/utils/RichRDDSuite.scala [hunk 6]
```diff
@@ -28,4 +28,28 @@ class RichRDDSuite extends SparkSuite {
...
+
+    val header = Array[Byte](108, 27, 1, 91)
+    val data = Array(Array[Byte](1, 19, 23, 127, -1), Array[Byte](23, 4, 15, -2, 1))
+    val r = sc.parallelize(data, numSlices = 2)
+    assert(r.getNumPartitions == 2)
+
+    val notParallelWrite = tmpDir.createTempFile("notParallelWrite")
+    r.saveFromByteArrays(notParallelWrite, tmpDir.tempDir, Some(header), parallelWrite = false)
+
+    assert(readBytes(notParallelWrite) sameElements header ++: data.flatten)
```
Comment (danking) [inline src/test/scala/is/hail/utils/RichRDDSuite.scala]: Since the types are the same, is `++:` any different from `++`? I think I prefer `++` because it's less unusual and people won't have to look up how `++:` differs.

File: src/main/scala/is/hail/utils/TempDir.scala (hunk 1 above)
Comment (jigold) [reply src/main/scala/is/hail/utils/TempDir.scala]: Needed to access the `tempDir` argument for the test suite. Changed it to `val localTempDir...`

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala (hunk 2 above)
Comment (jigold) [reply src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: I made it symmetric to `RichRDD.writeTable` which also overwrites by default. I can make a separate PR to change the behavior for both methods if needed.

File: src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala (hunk 4 above)
Comment (jigold) [reply src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala]: Correct.

File: src/test/scala/is/hail/utils/RichRDDSuite.scala (hunk 6 above)
Comment (jigold) [reply src/test/scala/is/hail/utils/RichRDDSuite.scala]: I get a compile error with `++` but not with `++:`.

File: src/main/scala/is/hail/utils/TempDir.scala [hunk 7]
```diff
@@ -38,7 +38,7 @@ object TempDir {
       createTempDir("/tmp", hConf))
 }
 
-class TempDir(localTempDir: String, tempDir: String) {
+class TempDir(val localTempDir: String, val tempDir: String) {
```
Comment (danking) [inline src/main/scala/is/hail/utils/TempDir.scala]: I think these are purposely hidden.

Even though it's called `createTempFile` it just creates a new filename thats not used. You can create a directory with that name if you want a directory instead of a file.

In this case, you can use `hc.tmpDir` since you're in a SparkSuite below.

File: src/test/scala/is/hail/utils/RichRDDSuite.scala (hunk 6 above)
Comment (danking) [reply src/test/scala/is/hail/utils/RichRDDSuite.scala]: I think you might need parentheses around it like:
```scala
assert(... sameElements (header ++ data.flatten))
```

Comment (danking) [reply src/test/scala/is/hail/utils/RichRDDSuite.scala]: The ++: is a right associative operator, so the compiler can figure out where to put the parentheses automatically.

I'm OK leaving it as is, I suppose if we start using `++:` more we'll start to recognize it more and it won't be so unusual.

**Comment by jigold** [comment]
I still couldn't get file to compile with "++" instead of "++:" even with parentheses.

**Comment by danking** [comment]
weird
